devito/ir/iet/scheduler.py

```diff
@@ -0,0 +1,158 @@
+        subs.update({i.expr: LoweredDimension(name=i.index.name, origin=i.expr)
+                     for i in uindices})
+    iet = SubstituteExpression(subs).visit(iet)
+
+    return iet
+
+
+def iet_make(clusters, dtype):
+    """
+    Create an Iteration/Expression tree (IET) given an iterable of :class:`Cluster`s.
+
+    :param clusters: The iterable :class:`Cluster`s for which the IET is built.
+    :param dtype: The data type of the scalar expressions.
+    """
+    processed = []
+    schedule = OrderedDict()
+    for cluster in clusters:
+        if not cluster.ispace.empty:
+            root = None
+            intervals = cluster.ispace.intervals
+
+            # Can I reuse any of the previously scheduled Iterations ?
+            index = 0
+            for i0, i1 in zip(intervals, list(schedule)):
+                if i0 != i1 or i0.dim in clusters.atomics[cluster]:
+                    break
+                root = schedule[i1]
+                index += 1
+            needed = intervals[index:]
+
+            # Build Iterations, including any necessary unbounded index
+            iters = []
+            for i in needed:
+                uindices = []
+                for j, offs in cluster.ispace.sub_iterators.get(i.dim, []):
+                    for n, o in enumerate(filter_ordered(offs)):
+                        name = "%s%d" % (j.name, n)
+                        vname = Scalar(name=name, dtype=np.int32)
+                        value = (i.dim + o) % j.modulo
+                        uindices.append(UnboundedIndex(vname, value, value, j, j + o))
+                iters.append(Iteration([], i.dim, i.dim.limits, offsets=i.limits,
+                                       uindices=uindices))
+
+            # Build Expressions
+            exprs = [Expression(v, np.int32 if cluster.trace.is_index(k) else dtype)
+                     for k, v in cluster.trace.items()]
+
+            # Compose Iterations and Expressions
+            body, tree = compose_nodes(iters + [exprs], retrieve=True)
+
+            # Update the current scheduling
+            scheduling = OrderedDict(zip(needed, tree))
+            if root is None:
+                processed.append(body)
+                schedule = scheduling
+            else:
+                nodes = list(root.nodes) + [body]
+                mapper = {root: root._rebuild(nodes, **root.args_frozen)}
+                transformer = Transformer(mapper)
+                processed = list(transformer.visit(processed))
+                schedule = OrderedDict(list(schedule.items())[:index] +
+                                       list(scheduling.items()))
+                for k, v in list(schedule.items()):
+                    schedule[k] = transformer.rebuilt.get(v, v)
+        else:
+            # No Iterations are needed
+            processed.extend([Expression(e, dtype) for e in cluster.exprs])
+
+    return List(body=processed)
+
+
+def iet_insert_C_decls(iet, func_table):
```

Comment:
note for reviewers: this is the former ``_insert_declarations``; nothing more, nothing less